pkg/cmd/extension/manager.go

```diff
@@ -417,58 +417,81 @@ func (m *Manager) installGit(cloneURL string, stdout, stderr io.Writer) error {
 }
 
 var localExtensionUpgradeError = errors.New("local extensions can not be upgraded")
+var upToDateError = errors.New("already up to date")
+var noExtensionsInstalledError = errors.New("no extensions installed")
 
 func (m *Manager) Upgrade(name string, force bool) error {
-	exe, err := m.lookPath("git")
-	if err != nil {
-		return err
-	}
-
-	exts := m.List(false)
+	// Fetch metadata during list only when upgrading all extensions.
+	// This is a performance improvement so that we don't make a
+	// bunch of unecessary network requests when trying to upgrade a single extension.
+	fetchMetadata := name == ""
+	exts, _ := m.list(fetchMetadata)
 	if len(exts) == 0 {
-		return errors.New("no extensions installed")
+		return noExtensionsInstalledError
+	}
+	if name == "" {
+		return m.upgradeExtensions(exts, force)
 	}
-
-	someUpgraded := false
 	for _, f := range exts {
-		if name == "" {
-			fmt.Fprintf(m.io.Out, "[%s]: ", f.Name())
-		} else if f.Name() != name {
+		if f.Name() != name {
 			continue
 		}
-
-		if f.IsLocal() {
-			if name == "" {
-				fmt.Fprintf(m.io.Out, "%s\n", localExtensionUpgradeError)
-			} else {
-				err = localExtensionUpgradeError
-			}
-			continue
+		var err error
+		// For single extensions manually retrieve latest version since we forgo
+		// doing it during list.
+		f.latestVersion, err = m.getLatestVersion(f)
+		if err != nil {
+			return err
 		}
+		return m.upgradeExtension(f, force)
+	}
+	return fmt.Errorf("no extension matched %q", name)
+}
 
-		binManifestPath := filepath.Join(filepath.Dir(f.Path()), manifestName)
-		if _, e := os.Stat(binManifestPath); e == nil {
-			err = m.upgradeBin(f)
-			someUpgraded = true
+func (m *Manager) upgradeExtensions(exts []Extension, force bool) error {
+	var failed bool
+	for _, f := range exts {
+		fmt.Fprintf(m.io.Out, "[%s]: ", f.Name())
+		err := m.upgradeExtension(f, force)
+		if err != nil {
+			if !errors.Is(err, localExtensionUpgradeError) &&
+				!errors.Is(err, upToDateError) {
```

Comment:
I like this change. But it seems unavailable at `gh version 2.3.0 (2021-12-02)`?

Reply:
Still I get a "already up to date" error, which I think should be OK and no more than a warning.

```
$ gh extension upgrade cherry-pick
X Failed upgrading extension cherry-pick: already up to date%   
```

Reply:
Will, they are in different code paths. Let me try to bring it even when you run `upgrade xxx`.